models/issue_milestone.go

```diff
@@ -311,71 +311,57 @@ func ChangeMilestoneStatus(m *Milestone, isClosed bool) (err error) {
 	return sess.Commit()
 }
 
-func changeMilestoneIssueStats(e *xorm.Session, issue *Issue) error {
-	if issue.MilestoneID == 0 {
-		return nil
-	}
-
-	m, err := getMilestoneByRepoID(e, issue.RepoID, issue.MilestoneID)
-	if err != nil {
-		return err
-	}
-
-	if issue.IsClosed {
-		m.NumOpenIssues--
-		m.NumClosedIssues++
-	} else {
-		m.NumOpenIssues++
-		m.NumClosedIssues--
-	}
+func updateMilestoneTotalNum(e Engine, milestoneID int64) (err error) {
+	_, err = e.Exec("UPDATE `milestone` SET num_issues=(SELECT count(*) FROM issue WHERE milestone_id=?) WHERE id=?",
+		milestoneID,
+		milestoneID,
+	)
+	return
+}
 
-	return updateMilestone(e, m)
+func updateMilestoneClosedNum(e Engine, milestoneID int64) (err error) {
+	_, err = e.Exec("UPDATE `milestone` SET num_closed_issues=(SELECT count(*) FROM issue WHERE milestone_id=? AND is_closed=?) WHERE id=?",
+		milestoneID,
+		true,
+		milestoneID,
+	)
+	return
 }
 
 func changeMilestoneAssign(e *xorm.Session, doer *User, issue *Issue, oldMilestoneID int64) error {
+	if err := updateIssueCols(e, issue, "milestone_id"); err != nil {
+		return err
+	}
+
 	if oldMilestoneID > 0 {
-		m, err := getMilestoneByRepoID(e, issue.RepoID, oldMilestoneID)
-		if err != nil {
+		if err := updateMilestoneTotalNum(e, oldMilestoneID); err != nil {
 			return err
 		}
-
-		m.NumIssues--
-		if issue.IsClosed {
-			m.NumClosedIssues--
-		}
-
-		if err = updateMilestone(e, m); err != nil {
+		if err := updateMilestoneClosedNum(e, oldMilestoneID); err != nil {
```

Comment:
It looks like this will execute always, whether the issue closed status has changed or not.

Reply:
No. The function `changeMilestoneAssign` will only be invoked when issues/pulls status changed.

Reply:
Isn't `changeMilestoneAssign()` used to move the target milestone of the issue? (e.g. 1.10.0 --> 1.11.0).
I've just set a breakpoint on the function and set a new milestone for one issue. The function was definitely called, and the issue never changed its `IsClosed` status.

Reply:
see line 336, this will be invoked only `if oldMilestoneID > 0 {`

Reply:
`oldMilestoneID > 0` != `isClosed`
I'm probably misunderstanding something but what I mean is:
* Milestone `1.10.0` has `ID`: 4; milestone `1.11.0` has `ID`: 5.
* Create issue `#123`; assign to milestone `1.10.0` (issue.MilestoneID=4).
    * `1.10.0` ==> num_issues = 1, num_closed_issues = 0
    * `1.11.0` ==> num_issues = 0, num_closed_issues = 0
* Edit issue `#123`; change milestone from `1.10.0` to `1.11.0`.

Now `changeMilestoneAssign()` is called with issue.MilestoneID=5, oldMilestoneID = 4, and issue.IsClosed = false. We expect:

* `1.10.0` ==> num_issues = 0, num_closed_issues = 0
* `1.11.0` ==> num_issues = 1, num_closed_issues = 0

In no case `num_closed_issues` should be updated, but `oldMilestoneID > 0` and `updateMilestoneClosedNum()` is called.